app/src/lib/stores/app-store.ts

```diff
@@ -2860,76 +2864,91 @@ export class AppStore extends TypedBaseStore<IAppState> {
     const { changesState, branchesState } = this.repositoryStateCache.get(
       repository
     )
-    const { tip } = branchesState
-    const currentBranch = tip.kind === TipState.Valid ? tip.branch : null
-    const hasChanges = changesState.workingDirectory.files.length > 0
-
-    if (
-      enableStashing() &&
-      hasChanges &&
-      currentBranch !== null &&
-      uncommittedChangesStrategy ===
-        UncommittedChangesStrategy.askForConfirmation
-    ) {
-      this._showPopup({
-        type: PopupType.StashAndSwitchBranch,
-        branchToCheckout: foundBranch,
-        repository,
-      })
 
-      return repository
-    }
+    let stashToPop: IStashEntry | null = null
+    if (enableStashing()) {
+      const hasChanges = changesState.workingDirectory.files.length > 0
+      if (
+        hasChanges &&
+        uncommittedChangesStrategy.kind ===
+          UncommittedChangesStrategyKind.askForConfirmation
+      ) {
+        this._showPopup({
+          type: PopupType.StashAndSwitchBranch,
+          branchToCheckout: foundBranch,
+          repository,
+        })
 
-    let shouldPopStash = false
+        return repository
+      }
 
-    if (enableStashing() && currentBranch !== null) {
+      const { tip } = branchesState
+      const currentBranch = tip.kind === TipState.Valid ? tip.branch : null
       if (
-        uncommittedChangesStrategy ===
-        UncommittedChangesStrategy.stashOnCurrentBranch
+        currentBranch !== null &&
+        uncommittedChangesStrategy.kind ===
+          UncommittedChangesStrategyKind.stashOnCurrentBranch
       ) {
-        await this.createStash(repository, currentBranch.name)
+        await this._createStashAndDropPreviousEntry(
+          repository,
+          currentBranch.name
+        )
       } else if (
-        uncommittedChangesStrategy ===
-        UncommittedChangesStrategy.moveToNewBranch
+        uncommittedChangesStrategy.kind ===
+        UncommittedChangesStrategyKind.moveToNewBranch
       ) {
-        await createDesktopStashEntry(repository, foundBranch.name)
-        shouldPopStash = true
+        const hasDeletedFiles = changesState.workingDirectory.files.some(
+          file => file.status.kind === AppFileStatusKind.Deleted
+        )
+        if (hasDeletedFiles) {
```

Comment:
Should this be

```suggestion
        if (
          hasDeletedFiles &&
          uncommittedChangesStrategy.transientStashEntry === null
        ) {
```

or is there a scenario where we want to go ahead with creating a new stash entry even if we've been passed one already?

Reply:
We shouldn't need this check because the working directory would be clean if there were a temporary stash entry at this point since the error handler calls `Dispatcher.moveChangesToBranchAndCheckout` which, after the wrapping the call on L5115 (called out in your comment below) would prevent the checkout from occurring. All that said, it doesn't hurt to have this additional safety check, so I'm going with it. 👍

Reply:
Addressed in 89b7437